### ISSUE TITLE
Can I set environment variables for the xunit runner?

### ISSUE THREAD

**Issue by dagood** [dotnet/arcade#3077]
Core-Setup's current tests rely on env vars being set to use during the tests, but I don't see any way to pass them in with Xunit.targets:

https://github.com/dotnet/arcade/blob/dc538a29793fd56618d0fa3186e2388d47d00c19/src/Microsoft.DotNet.Arcade.Sdk/tools/XUnit/XUnit.targets#L89-L91

I see a few other projects seem to have gone through some effort to derive info from the basics that are passed in:

https://github.com/dotnet/core-sdk/blame/master/test/Microsoft.DotNet.Tools.Tests.Utilities/RepoDirectoriesProvider.cs ("enable tests" commit, this is specifically during the migration to Arcade)

https://github.com/dotnet/sdk/blob/master/src/Tests/Microsoft.NET.TestFramework/TestContext.cs

Maybe that's the way to go and asking for env vars is a nonstarter, but I wanted to ask. I'm not very familiar with the tests in Core-Setup and would have liked to do a straight port for now.

Is it reasonable to ask for the ability to define env vars to pass to the xunit runner?

@markwilkie

**Comment by alexperovich** [comment]
What is the scope of the environment variables? Can they just be set in the build definition?

**Comment by dagood** [comment]
They are to pass in directories, product version (to restore built packages), and the test target arch/configuration/RID in this case:

https://github.com/dotnet/core-setup/blob/0298b805eaa286192865be92ec351e7a74b177af/src/test/dir.proj#L94

This applies to local builds.

**Comment by dagood** [comment]
For now I've overridden the RunTests target to add support:

```xml
<Exec
  Command='$(_TestRunnerCommand)'
  EnvironmentVariables="@(TestRunnerEnvironmentVariable)"
  LogStandardErrorAsError="false"
  WorkingDirectory="$(_TargetDir)"
  IgnoreExitCode="true">
```

(I'm not very close to merging it into the mainline repo though, so plenty of time for better advice. 😄)

**Comment by alexperovich** [comment]
I don't have a problem with environment variables, but whatever the solution is needs to support VS test explorer. It should be easy to run all the tests in VS.

**Comment by dagood** [comment]
On the flip side, if it turns out environment variables simply can't work from the VS test explorer, that sounds like a good reason to me not to accept this in the Arcade SDK. (I haven't worked on the Windows built yet so I can't check now.)

**Comment by tmat** [comment]
It would be better if the tests themselves read the information from somewhere else, not from environment. One option might be to generate a text file with these values during the build, copy it to the output dir and then read that info from the tests.

**Comment by dagood** [comment]
Not knowing anything about test runners, why is that better?

**Comment by tmat** [comment]
Because it just works regardless of where the test is run from (VS/command line).

**Comment by ViktorHofer** [comment]
> Because it just works regardless of where the test is run from (VS/command line).

VS Test Explorer doesn't use Arcade's test target therefore I don't understand how this would work in VS?

**Comment by tmat** [comment]
@ViktorHofer Regardless of what runner is used, tests depending on environment variables only work in an environment where these variables are set. If you launch VS directly (as opposed to some pre-configured developer command prompt) it won't have the variables set.

**Comment by dagood** [comment]
The trick here is that *building* the tests in VS *does* use Arcade, so we'd have access to any props we need at that point in time to generate the config file, right?

**Comment by tmat** [comment]
Correct.

**Comment by dagood** [comment]
Sounds reasonable to me. Thanks. 😄

**Comment by tmat** [comment]
One thing to watch for is to avoid including machine specific information in the generate config file. The tests might run on a different machine than the build did.

**Comment by dagood** [comment]
I implemented the var transfer file in Core-Setup, by the way, and it works great. There was previously a `visual-studio-devenv.cmd` script that set up env vars and launched VS, and it isn't necessary anymore. (I wasn't actually aware of the script when I opened this issue.)

The file does contain machine-specific information, because that's what was specifically needed by the tests. But I left in env var override capabilities for any odd situations that come up in the future.
